hr_timesheet_overview/models/hr_employee_hour.py

```diff
@@ -0,0 +1,327 @@
+        )
+        # We only take valid days of work
+        if attendances and (not global_leaves or exclude_global_leaves):
+            hours_qty, days_qty = self.get_time_quantities_from_calendar(
+                calendar, attendances
+            )
+            values = {
+                "name": contract.name,
+                "model_name": "hr.contract",
+                "res_id": contract.id,
+                "date": date,
+                "employee_id": contract.employee_id.id,
+                "type": "contract",
+                "hours_qty": -hours_qty,
+                "days_qty": -days_qty,
+                "sum_type": "remove",
+            }
+        return values
+
+    @api.model
+    def _get_attendances_values(self, employee, date_from):
+        """Retrieve attendances (from resource module) values mapped by date
+
+        These attendances are the matrix product of contractual hours by day
+        and week days (NOT the `hr.attendance` model).
+
+        Warning: it is intended to be injected in context to reduce calls
+
+        :param employee: an employee record
+        :param date_from: a datetime.date object (default first contract date, included)
+        :param date_to: a datetime.date object (default today, included)
+        """
+        date_to = fields.Date.today()
+        values_by_date = {}
+        for contract in employee.contract_ids:
+            _logger.debug(
+                f"""process contract '{contract.name}'
+                from {contract.last_hours_report_date} to {date_to}"""
+            )
+            ranged_dates = generate_dates_from_range(
+                contract.last_hours_report_date, date_to
+            )
+            for date in ranged_dates:
+                if date_from <= date >= date_to:
+                    continue
+                values = self._prepare_attendance_value(contract, date)
+                if values:
+                    values_by_date[date] = values
+        return list(values_by_date.values())
+
+    @api.model
+    def _hook_get_type_from_timesheet(self, timesheet):
+        """Hook to allow proper override of type value
+        Called in `_prepare_timesheets_lines` method
+        """
+        return "leave" if timesheet.holiday_id else "timesheet"
+
+    @api.model
+    def _hook_get_name_from_timesheet(self, timesheet):
+        """Hook to allow proper override of name value
+        Called in `_prepare_timesheets_lines` method
+        """
+        return timesheet.account_id.name if timesheet.holiday_id else timesheet.name
+
+    def _prepare_timesheet_line(self, timesheet):
+        """Retrieve a timesheet line values
+
+        Day ratio quantity is a percentage of consumed hours relative to the
+        attendance day max quantity to make a full day (2h on 8h day is 25%).
+
+        If `attendances_by_date` is not set in context or date is not found,
+        default hours and day quantity are used (cf: `DEFAULT_TIME_QTY`)
```

Comment:
can you explain the functional side of the trick of processing timesheet records with attendance at the same time 
do we need to specifically mark records from the timesheet created out of working hours?

Reply:
You need to know the correct contractual hours of one particular timesheet date, to know what is the day percentage of that timesheet amount regarding the correct contract.
For this you need to access the contract and process the days defined in the attendances to know how much.
To do so and also to avoid too much read requests and relation joins, i process the attendances in one batch and then propagate a dict containing hours length of day and day ratio in percentage to ease retrieval.

Example:
 - Actual contract since 01/02/2022 (Mon to Fr) with 7.33h daily except wednesday afternoon
 - Previous contrat from 01/02/2021 to 31/01/2022 idem but with 8h daily and wednesday afternoon included
 - Resource linked to the employee is the actual contract.

A timesheet for the 02/06/2021 is 2.5h amount and is a wednesday.
 - the daily hours to do is 8h for this date (previous contract) then the timesheet days qty should be 1 day
 - if you take the linked ressource then daily hours is 3,665 then the timesheet is in fact ~0,68 day

Result : 0.32 days of difference (with hundred even thousands of timesheets, the results are really really wrong)